File: Java/프로그래머스/2/159993. 미로 탈출/미로 탈출.java
import java.util.*;

class Solution {
    public int solution(String[] maps) {
        int answer = -1;
        int[] start = {0, 0};
        int[] end = {0, 0};
        int[] lever = {0, 0};
        int n = maps.length;
        int m = maps[0].length();

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                char c = maps[i].charAt(j);
                if (c == 'S') {
                    start[0] = i;
                    start[1] = j;
                } else if (c == 'E') {
                    end[0] = i;
                    end[1] = j;
                } else if (c == 'L') {
                    lever[0] = i;
                    lever[1] = j;
                }
            }
        }

        int a = bfs(maps, start, lever);
        int b = bfs(maps, lever, end);
        if (a != -1 && b != -1) {
            answer = a + b;
        }
        return answer;
    }

    int bfs(String[] maps, int[] start, int[] end) {
        int n = maps.length;
        int m = maps[0].length();
        int[] dx = {0, 0, 1, -1};
        int[] dy = {1, -1, 0, 0};
        Queue<int[]> queue = new LinkedList<>();
        int[][] visited = new int[n][m];
        int x = start[0];
        int y = start[1];

        queue.add(new int[]{x, y});

        while (!queue.isEmpty()) {
            int[] now = queue.poll();
            x = now[0];
            y = now[1];

            if (x == end[0] && y == end[1]) {
                return visited[x][y];
            }

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                if (nx < 0 || nx >= n || ny < 0 || ny >= m || maps[nx].charAt(ny) == 'X' || visited[nx][ny] > 0) {
                    continue;
                }
                visited[nx][ny] = visited[x][y] + 1;
                queue.add(new int[]{nx, ny});
            }
        }
        return -1;
    }
}

File: Python/프로그래머스/2/159993. 미로 탈출/미로 탈출.py
def solution(maps):
    start = None
    lever = None
    end = None

    for i in range(len(maps)):
        for j in range(len(maps[i])):
            if maps[i][j] == "S":
                start = (i, j)
            if maps[i][j] == "L":
                lever = (i, j)
            if maps[i][j] == "E":
                end = (i, j)

    start_lever = bfs(start, lever, maps)
    if start_lever == -1:
        return -1

    lever_end = bfs(lever, end, maps)
    if lever_end == -1:
        return -1

    return start_lever + lever_end


def bfs(start, target, maps):
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    visited = [[False] * (len(maps[0])) for _ in range(len(maps))]
    visited[start[0]][start[1]] = True
    q = [(start[0], start[1], 0)]

    while q:
        x, y, dist = q.pop(0)

        if (x, y) == target:
            return dist

        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if (
                0 <= nx < len(maps)
                and 0 <= ny < len(maps[0])
                and maps[nx][ny] != "X"
                and not visited[nx][ny]
            ):
                visited[nx][ny] = True
                q.append((nx, ny, dist + 1))

    return -1